miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+
+    TODO: use u32, i16, etc. for format.
+    """
+
+    def _pack(self, number):
+        return super(Num, self)._pack([number])
+
+    def _unpack(self, raw_str):
+        upck = super(Num, self)._unpack(raw_str)
+        if len(upck) > 1:
+            raise ValueError("Num format string unpacks to multiple values, "
+                             "should be 1")
+        return upck[0]
+
+
+class Ptr(Num):
+    """Special case of number of which value indicates the address of a
+    MemStruct. Provides deref_<field> as well as <field> when used, to set and
+    get the pointed MemStruct.
+    """
+
+    def __init__(self, fmt, dst_type, *type_args, **type_kwargs):
+        """Args:
+            fmt: (str) Num compatible format that will be the Ptr representation
+                in memory
+            dst_type: (MemStruct or MemField) the MemStruct this Ptr points to.
+                If a MemField is given, it is transformed into a MemStruct with
+                mem(TheMemField).
+            *type_args, **type_kwargs: arguments to pass to the the pointed
+                MemStruct when instanciating it (e.g. for MemStr encoding or
+                MemArray field_type).
+        """
+        if not isinstance(dst_type, MemField) and\
+                not (isinstance(dst_type, type) and\
+                        issubclass(dst_type, MemStruct)) and\
+                not dst_type == MemSelf:
+            raise ValueError("dst_type of Ptr must be a MemStruct type, a "
+                             "MemField instance, the MemSelf marker or a class "
+                             "name.")
+        super(Ptr, self).__init__(fmt)
+        if isinstance(dst_type, MemField):
+            # Patch the field to propagate the MemSelf replacement
+            dst_type._get_self_type = lambda: self._get_self_type()
+            dst_type = mem(dst_type)
+        self._dst_type = dst_type
+        self._type_args = type_args
+        self._type_kwargs = type_kwargs
+
+    def _fix_dst_type(self):
+        if self._dst_type == MemSelf:
+            if self._get_self_type() is not None:
+                self._dst_type = self._get_self_type()
+            else:
+                raise ValueError("Unsupported usecase for MemSelf, sorry")
+
+    @property
+    def dst_type(self):
+        """Return the type (MemStruct subtype) this Ptr points to."""
+        self._fix_dst_type()
+        return self._dst_type
+
+    def deref_get(self, vm, addr):
+        """Deserializes the data in @vm (VmMngr) at @addr to self.dst_type.
+        Equivalent to a pointer dereference rvalue in C.
+        """
+        return self.dst_type(vm, addr, *self._type_args, **self._type_kwargs)
+
+    def deref_set(self, vm, addr, val):
+        """Serializes the @val MemStruct subclass instance in @vm (VmMngr) at
+        @addr. Equivalent to a pointer dereference assignment in C.
+        """
+        # Sanity check
+        if self.dst_type != val.__class__:
```

Comment:
This kind of thing, in a C or C++ compiler, would raise a warning or an error. The idea is that if you subclass you `MemStruct` to add fields, an assignment of the subclass to the parent _should_ raise a warning. If you subclass it just to add methods, it's okay, but I do not see real usecases for that.

In both cases, casting your subclass to the superclass (via the `.cast(<superclass>)`) method, as you would do in C, is the way to go if you want to get rid of the warning. This is the way to indicate that you know what you are doing. Is this explanation ok for you?